File: backend/dotnet/nlayer-architecture/full/src/App.Template.Api/Controllers/ExportController.cs
using App.Template.Api.Models.Common;
using App.Template.Api.Models.Dtos;
using App.Template.Api.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace App.Template.Api.Controllers;

[Authorize]
[ApiController]
[Route("api/export")]
public class ExportController : ControllerBase
{
    private readonly IExportService _exportService;
    private readonly IUserService _userService;
    private readonly IDepartmentService _departmentService;
    private readonly IAuditLogService _auditLogService;
    private readonly INotificationService _notificationService;

    public ExportController(
        IExportService exportService,
        IUserService userService,
        IDepartmentService departmentService,
        IAuditLogService auditLogService,
        INotificationService notificationService)
    {
        _exportService = exportService;
        _userService = userService;
        _departmentService = departmentService;
        _auditLogService = auditLogService;
        _notificationService = notificationService;
    }

    [HttpGet("users")]
    public async Task<IActionResult> ExportUsers(
        [FromQuery] string format = "xlsx",
        [FromQuery] string? search = null,
        [FromQuery] bool? isActive = null,
        [FromQuery] long? departmentId = null)
    {
        var query = new UsersQueryParams { Page = 1, PageSize = int.MaxValue, Search = search, IsActive = isActive, DepartmentId = departmentId };
        var result = await _userService.GetUsersAsync(query);
        return await ExportData(result.Items, format, "users");
    }

    [HttpGet("departments")]
    public async Task<IActionResult> ExportDepartments(
        [FromQuery] string format = "xlsx",
        [FromQuery] string? search = null,
        [FromQuery] bool? isActive = null)
    {
        var query = new DeptQueryParams { Page = 1, PageSize = int.MaxValue, Search = search, IsActive = isActive };
        var result = await _departmentService.GetDepartmentsAsync(query);
        return await ExportData(result.Items, format, "departments");
    }

    [HttpGet("audit-logs")]
    public async Task<IActionResult> ExportAuditLogs(
        [FromQuery] string format = "xlsx",
        [FromQuery] string? entityType = null,
        [FromQuery] string? action = null,
        [FromQuery] DateTime? fromDate = null,
        [FromQuery] DateTime? toDate = null)
    {
        var query = new AuditLogsQueryParams { Page = 1, PageSize = int.MaxValue, EntityType = entityType, Action = action, FromDate = fromDate, ToDate = toDate };
        var result = await _auditLogService.GetAuditLogsAsync(query);
        return await ExportData(result.Items, format, "audit-logs");
    }

    [HttpGet("notifications")]
    public async Task<IActionResult> ExportNotifications([FromQuery] string format = "xlsx")
    {
        var userIdStr = User.FindFirst(System.Security.Claims.ClaimTypes.NameIdentifier)?.Value ?? "0";
        var userId = long.Parse(userIdStr);
        var query = new NotificationsQueryParams { Page = 1, PageSize = int.MaxValue };
        var result = await _notificationService.GetNotificationsAsync(userId, query);
        return await ExportData(result.Items, format, "notifications");
    }

    private async Task<IActionResult> ExportData<T>(IEnumerable<T> data, string format, string entityName)
    {
        var timestamp = DateTime.UtcNow.ToString("yyyyMMdd");
        return format.ToLower() switch
        {
            "csv" => File(await _exportService.ExportToCsvAsync(data), "text/csv", $"{entityName}_{timestamp}.csv"),
            "pdf" => File(await _exportService.ExportToPdfAsync(data, $"{entityName} Export"), "application/pdf", $"{entityName}_{timestamp}.pdf"),
            _ => File(await _exportService.ExportToExcelAsync(data, entityName), "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet", $"{entityName}_{timestamp}.xlsx")
        };
    }
}

File: backend/nestjs/feature-architecture/full/src/features/export/export.controller.ts
import { Controller, Get, Query, Res, UseGuards, Request } from '@nestjs/common';
import { Response } from 'express';
import { AuthGuard } from '@nestjs/passport';
import { ApiTags, ApiOperation, ApiResponse, ApiBearerAuth } from '@nestjs/swagger';
import { ExportService } from './export.service';
import { SkipTransform } from '../../common/decorators/response-message.decorator';

@ApiTags('Export')
@Controller('export')
@UseGuards(AuthGuard('jwt'))
@ApiBearerAuth('JWT-auth')
export class ExportController {
  constructor(private readonly exportService: ExportService) {}

  @Get('users')
  @SkipTransform()
  @ApiOperation({ summary: 'Export users' })
  @ApiResponse({ status: 200, description: 'File download' })
  async exportUsers(
    @Query('format') format: string = 'xlsx',
    @Query('search') search?: string,
    @Query('isActive') isActive?: string,
    @Query('departmentId') departmentId?: string,
    @Res() res?: Response,
  ) {
    const { buffer, contentType, fileName } = await this.exportService.exportUsers(format, search, isActive, departmentId);
    res!.set({
      'Content-Type': contentType,
      'Content-Disposition': `attachment; filename="${fileName}"`,
      'Content-Length': buffer.length,
    });
    res!.end(buffer);
  }

  @Get('departments')
  @SkipTransform()
  @ApiOperation({ summary: 'Export departments' })
  @ApiResponse({ status: 200, description: 'File download' })
  async exportDepartments(
    @Query('format') format: string = 'xlsx',
    @Query('search') search?: string,
    @Query('isActive') isActive?: string,
    @Res() res?: Response,
  ) {
    const { buffer, contentType, fileName } = await this.exportService.exportDepartments(format, search, isActive);
    res!.set({
      'Content-Type': contentType,
      'Content-Disposition': `attachment; filename="${fileName}"`,
      'Content-Length': buffer.length,
    });
    res!.end(buffer);
  }

  @Get('audit-logs')
  @SkipTransform()
  @ApiOperation({ summary: 'Export audit logs' })
  @ApiResponse({ status: 200, description: 'File download' })
  async exportAuditLogs(
    @Query('format') format: string = 'xlsx',
    @Query('entityType') entityType?: string,
    @Query('action') action?: string,
    @Query('fromDate') fromDate?: string,
    @Query('toDate') toDate?: string,
    @Res() res?: Response,
  ) {
    const { buffer, contentType, fileName } = await this.exportService.exportAuditLogs(format, entityType, action, fromDate, toDate);
    res!.set({
      'Content-Type': contentType,
      'Content-Disposition': `attachment; filename="${fileName}"`,
      'Content-Length': buffer.length,
    });
    res!.end(buffer);
  }

  @Get('notifications')
  @SkipTransform()
  @ApiOperation({ summary: 'Export notifications' })
  @ApiResponse({ status: 200, description: 'File download' })
  async exportNotifications(
    @Query('format') format: string = 'xlsx',
    @Request() req?: any,
    @Res() res?: Response,
  ) {
    const { buffer, contentType, fileName } = await this.exportService.exportNotifications(format, req!.user.userId);
    res!.set({
      'Content-Type': contentType,
      'Content-Disposition': `attachment; filename="${fileName}"`,
      'Content-Length': buffer.length,
    });
    res!.end(buffer);
  }
}
